subsys/instrumentation/ringbuffer/ringbuffer.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/instrumentation/instrumentation_ringbuffer.h>
+
+#define RING_BUFFER_SIZE CONFIG_INSTRUMENTATION_MODE_CALLGRAPH_TRACE_BUFFER_SIZE
+static uint8_t ring_buffer[RING_BUFFER_SIZE];
+
+static int head;
+static int tail;
+static int wrapped;
+
+static size_t type_size_table[INSTR_EVENT_NUM];
+
+__no_instrumentation__
+size_t instr_rb_get_item_size(enum instr_event_types type)
+{
+	return type_size_table[type];
+}
+
+__no_instrumentation__
+int instr_rb_increment_pos(int pos, int offset)
+{
+	pos += offset;
+	pos %= RING_BUFFER_SIZE;
+
+	return pos;
+}
+
+__no_instrumentation__
+struct instr_record *instr_rb_put_item_claim(enum instr_event_types item_type)
+{
+	int item_size;
+	int new_head;
+	int tail_record_size;
+	struct instr_record *tail_record;
+
+	item_size = instr_rb_get_item_size(item_type);
+	new_head = head + item_size;
+	/* If there is space to put item, adjust tail accordingly. */
+	if ((head + item_size) <= RING_BUFFER_SIZE) {
+		if (head == tail && wrapped) {
+			tail_record = (struct instr_record *)&ring_buffer[tail];
+			tail_record_size = instr_rb_get_item_size(tail_record->header.type);
+			tail = instr_rb_increment_pos(tail, tail_record_size);
+		}
+
+		while (head < tail && new_head > tail) {
+			tail_record = (struct instr_record *)&ring_buffer[tail];
+			tail_record_size = instr_rb_get_item_size(tail_record->header.type);
+			tail = instr_rb_increment_pos(tail, tail_record_size);
+		}
+
+		return (struct instr_record *)&ring_buffer[head];
+	}
+
+	/* No space to put item, buffer will wrap. */
+	return NULL;
+}
+
+__no_instrumentation__
+int instr_rb_put_item_wrapping(struct instr_record *rec)
+{
+	int new_head;
+	int next_tail;
+	struct instr_record *tail_record;
+	size_t first_chunk;
+	size_t second_chunk;
+
+	assert(rec != NULL);
+
+	/* Adjust head first */
+	new_head = instr_rb_increment_pos(head, instr_rb_get_item_size(rec->header.type));
+
+	/*
+	 * Handle case when tail points to an item which is wrapped itself,
+	 * like, for instance, when buffer size = 41, head = 39, tail = 40, and
+	 * a new item of size 20 will be inserted.
+	 */
+	tail_record = (struct instr_record *)&ring_buffer[tail];
+	next_tail = instr_rb_increment_pos(tail, instr_rb_get_item_size(tail_record->header.type));
+	if (next_tail < tail) {
+		tail = next_tail;
+	}
+
+	/*
+	 * Advance tail until there is space to put the new item, but not beyond
+	 * the head.
+	 */
+	while (new_head > tail && tail != head) {
+		tail_record = (struct instr_record *)&ring_buffer[tail];
+		tail = instr_rb_increment_pos(tail,
+						instr_rb_get_item_size(tail_record->header.type));
+	}
+
+	/* First, write the number of bytes available before buffer wraps. */
+	first_chunk = RING_BUFFER_SIZE - head;
+	memcpy(&ring_buffer[head], (uint8_t *)rec, first_chunk);
+
+	/* Wrap buffer. */
+	head = 0;
+
+	/* Second, write remaining item bytes not written yet. */
```

Comment:
Just a query on these, assuming that this is running in a multi-core system, how does this protect against 2 cores trying to do things with the ring buffer at once?